### PR TITLE
Adds /home/USER/.local/bin/ to PATH in /etc/sudoers.d/vscode,  

### PR DESCRIPTION
**Pull request by prabhakk-mw**
fixes devcontainers/features#870

### CODE REVIEW

File: src/python/install.sh
Comment (samruddhikhandale) [inline src/python/install.sh:783]: Can you bump the patch version of the Feature? Thanks! 

https://github.com/devcontainers/features/blob/4ccebdd6bc93110c4e7c1c8a5cb2f3c9df706cc7/src/python/devcontainer-feature.json#L3

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
One minor comment, besides that looks good to me!

**Comment by samruddhikhandale** [comment]
A friendly ping @prabhakk-mw, it would be wonderful to merge this PR in, can you help us address one minor comment? Thanks!

File: src/python/install.sh
Comment (samruddhikhandale) [inline src/python/install.sh:871]: Rethinking about this one, if you have multiple `Defaults secure_path` lines under /etc/sudoers.d/, the last one encountered will be the one that takes effect. This is because each `Defaults secure_path` line encountered will overwrite the previous one.

Hence, if you want to add to the `secure_path`, you need to check if this entry exists, is so then you would need to first extract the current `secure_path`, add your new path to it, and then write it back.

Something like this 👇 ( this is not tested, so would be good to test it)

```
# Define the new path
new_path="/home/${USERNAME}/.local/bin"

# Check if the file contains "Defaults secure_path"
if sudo grep -q "Defaults secure_path" /etc/sudoers.d/$USERNAME; then
  # Extract the current secure_path
  current_path=$(sudo grep secure_path /etc/sudoers.d/$USERNAME | sed -e 's/Defaults secure_path="//' -e 's/"//')

  # Add the new path
  secure_path="${current_path}:${new_path}"

  # Replace the secure_path in the sudoers file
  sudo sed -i -e "s|Defaults secure_path=\"${current_path}\"|Defaults secure_path=\"${secure_path}\"|" /etc/sudoers.d/$USERNAME
else
  # If "Defaults secure_path" doesn't exist in the file, append it
  echo "Defaults secure_path=\"${new_path}:${PATH}\"" >> /etc/sudoers.d/$USERNAME
fi
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]
Let one comment, thanks!

**Comment by prabhakk-mw** [comment]
> 
> @prabhakk-mw please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > - (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> >```
> >@microsoft-github-policy-service agree
> >```
> > - (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> >```
> >@microsoft-github-policy-service agree company="Microsoft"
> >```
> 
> <details>
>     <summary>Contributor License Agreement</summary>
>     
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**),
> and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
> contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
> date below.
> 1. **Definitions**.
> **“Code”** means the computer software code, whether in human-readable or machine-executable form,
> that is delivered by You to Microsoft under this Agreement.
> **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
> approved by the Open Source Initiative (www.opensource.org).
> **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
> Project, including but not limited to communication on electronic mailing lists, source code control
> systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
> discussing and improving that Project, but excluding communication that is conspicuously marked or
> otherwise designated in writing by You as “Not a Submission.”
> **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
> associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
> Project. This Agreement covers any and all Submissions that You, now or in the future (except as
> described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
> Should You wish to Submit materials that are not Your original work, You may Submit them separately
> to the Project if You (a) retain all copyright and license information that was in the materials as You
> received them, (b) in the description accompanying Your Submission, include the phrase “Submission
> containing materials of a third party:” followed by the names of the third party and any licenses or other
> restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
> guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
> for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
> Submission is made in the course of Your work for an employer or Your employer has intellectual
> property rights in Your Submission by contract or applicable law, You must secure permission from Your
> employer to make the Submission before signing this Agreement. In that case, the term “You” in this
> Agreement will refer to You and the employer collectively. If You change employers in the future and
> desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
> and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
>   * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
> indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
> Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
> the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
> parties.
>   * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
> indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
> Your patent claims that are necessarily infringed by the Submission or the combination of the
> Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
> import or otherwise dispose of the Submission alone or with the Project.
>   * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
> No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
> granted by implication, exhaustion, estoppel or otherwise.
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
> licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
> have disclosed under Section 3). You represent that You have secured permission from Your employer to
> make the Submission in cases where Your Submission is made in the course of Your work for Your
> employer or Your employer has intellectual property rights in Your Submission by contract or applicable
> law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
> have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
> You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
> REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
> EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
> PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
> NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
> You later become aware that would make Your representations in this Agreement inaccurate in any
> respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
> contributions may be maintained indefinitely and disclosed publicly, including Your name and other
> information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
> the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
> Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
> exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
> defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
> supersedes any and all prior agreements, understandings or communications, written or oral, between
> the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.
> 
> </details>
> 

@microsoft-github-policy-service agree company="MathWorks"

**Comment by samruddhikhandale** [comment]
@prabhakk-mw Hi 👋 
Curious if you got the chance to address https://github.com/devcontainers/features/pull/887#discussion_r1526676190 
Let us know if you are busy and if you'd prefer then we can get someone to complete the PR on your behalf.

**Comment by prabhakk-mw** [comment]
Hi @samruddhikhandale 

Thank you for circling back, I'll make time to push your suggested changes this week.

**Comment by prabhakk-mw** [comment]
@samruddhikhandale 
Please have a look at the latest changes.
I'll work on the tests on Monday.

**Comment by prabhakk-mw** [comment]
@samruddhikhandale 
Added additional tests and bumped up the version number to 1.4.4

**Review by samruddhikhandale — COMMENTED** [review]
Left some minor comments, thanks!

File: src/python/install.sh
Comment (prabhakk-mw) [reply src/python/install.sh:875]: @samruddhikhandale how about `grep -qs  ${SEARCH_STR} ${SUDOERS_FILE}; then` instead:

man pages from grep:
```
      -q, --quiet, --silent
              Quiet; do not write anything to standard output.  Exit immediately with zero status if any match is found, even if an error was detected.  Also see the -s or --no-messages option.

       -s, --no-messages
              Suppress error messages about nonexistent or unreadable files.

```

Comment (prabhakk-mw) [inline src/python/install.sh:875]:
```suggestion
        if grep -qs ${SEARCH_STR} ${SUDOERS_FILE}; then
```

Comment (samruddhikhandale) [reply src/python/install.sh]: Yep, makes sense to me! Waiting for the test to pass and we can merge it ⚡

**Comment by samruddhikhandale** [comment]
The tests are failing even on the `main` branch, hence, we cannot merge this PR until we fix that. @gauravsaini04 Can you help prioritize fixing the tests?

Also, it might be worth checking if updating [action-runner](https://github.com/devcontainers/features/blob/main/.github/workflows/test-all.yaml#L62) to `devcontainer-image-builder-ubuntu` could help.

**Comment by prabhakk-mw** [comment]
Hi @samruddhikhandale 

Are we still waiting for the fixes to the main branch?
Or was there something I needed to do to unblock this. 

Thanks

**Comment by samruddhikhandale** [comment]
@prabhakk-mw Thanks for the nudge! Can you update https://github.com/devcontainers/features/blob/main/.github/workflows/test-all.yaml#L63 to `devcontainer-image-builder-ubuntu` in your PR? Let's see if it fixes your test.

**Comment by prabhakk-mw** [comment]
The logs dont seem to suggest that the test is failing due to the change in this Pull Request:




Am I reading it wrong?-

**Comment by samruddhikhandale** [comment]
Yep, however as a standard practice we'd like to make sure the tests are green before we merge in any new PRs.

**Comment by samruddhikhandale** [comment]
@gauravsaini04 Can you prioritize fixing the tests? thanks!

**Comment by prabhakk-mw** [comment]
Hi @samruddhikhandale @gauravsaini04 

The reason the CENTOS7 scenario is failing seems to be because of incorrectly configured PATHS for the openssl installation. Reverting to using the default installation paths for it based on their [doc](https://github.com/openssl/openssl/blob/master/INSTALL.md#building-openssl) seems to have fixed it.

A potential fix for the failing `centos7` build is:

```diff
diff --git a/src/python/install.sh b/src/python/install.sh
index 1aad3f6..f3eea3d 100755
--- a/src/python/install.sh
+++ b/src/python/install.sh
@@ -390,7 +390,7 @@ add_symlink() {
 }
 
 install_openssl3() {
-    local _prefix=$1
+    # local _prefix=$1
     mkdir /tmp/openssl3
     (
         cd /tmp/openssl3
@@ -403,7 +403,8 @@ install_openssl3() {
         curl -sSL -o "/tmp/openssl3/${tgz_filename}" "${tgz_url}"
         tar xzf ${tgz_filename}
         cd openssl-${openssl3_version}
-        ./config --prefix=${_prefix} --openssldir=${_prefix} --libdir=lib
+        # ./config --prefix=${_prefix} --openssldir=${_prefix} --libdir=lib
+        ./config --libdir=lib
         make -j $(nproc)
         make install_dev
     )
@@ -446,11 +447,12 @@ install_from_source() {
     # Some platforms/os versions need modern versions of openssl installed
     # via common package repositories, for now rhel-7 family, use case statement to
     # make it easy to expand
+    INSTALL_PATH="/usr"
     case ${VERSION_CODENAME} in
         centos7|rhel7)
             check_packages perl-IPC-Cmd
-            install_openssl3 ${INSTALL_PATH}
-            ADDL_CONFIG_ARGS="--with-openssl=${INSTALL_PATH} --with-openssl-rpath=${INSTALL_PATH}/lib"
+            install_openssl3
+            ADDL_CONFIG_ARGS="--with-openssl=${INSTALL_PATH}/local --with-openssl-rpath=${INSTALL_PATH}/local/lib"
             ;;
     esac
 

```
With these changes I was able to get the test passing:
`devcontainer features test -f python --filter centos-7 --skip-autogenerated `

Result:
```
Test Passed!
🧪 Executing duplicate test for feature 'python'...
⚠️ Skipping duplicate test for python because '/mathworks/devel/sandbox/prabhakk/cit/codespaces/features/test/python/duplicate.sh' does not exist.
🧹 Cleaning up 1 test containers...
🧹 Removing container 119945e980df...



  ================== TEST REPORT ==================
✅ Passed:      'centos-7'
```

File: src/python/devcontainer-feature.json
Comment (samruddhikhandale) [inline src/python/devcontainer-feature.json:3]:
```suggestion
  "version": "1.6.0",
```

**Review by Thompson1985 — APPROVED** [review]

**Pull request merged**
